### DISCUSSION TITLE
Welcome to mySpot Discussions!

### DISCUSSION THREAD

**Discussion by isaacpasc** [isaacpasc/mySpot#4] [Announcements]
## 👋 Welcome!
  This code base is in dire need of help, please:
  * Ask any questions or doubts about todo items or understanding the code
  * Share ideas
  * Share any known issues

**Comment by Ambiaence** [comment]
Glad to see you open this up. Will the potential android version also be open source? 

Just an idea. Perhaps the framework you use for android could eventually be used for both IOS and Android to make development easier.

This is one heck of a first project Isaac :thumbsup:. I really am looking forward to following and using this project.

**Reply by isaacpasc** [reply]
I would love to learn react native and this could be used for both ios and android but for now this project was only meant to teach me native iOS development and it grew a little more than expected. I definitely want to also learn native android development as well. And yes all my projects will be open source! And I appreciate the support!

**Comment by ramonteiro** [comment]
It's a great job for a first project! Congrats!! I definitely agree to localizate the app even with an automatic process like Google, because if the project is cool some native colleague will translate its strings. 

Understood your point about view model. I've just tested it and it seems possible to set the subviews in the observable class/VM, but a better choice is only removing the @State vars from views then declare them as @Published in a new file. Let's try clean up some views. To isolate the sub views and make their main views smaller, maybe the factory design pattern could work for that, let's see. 

About the CloudKit... it's a great native tool but I have to check deeply the issues you said in the project. I have only one project with it well implemented (much pain to work well😅) maybe a more experienced colleague can help you in this point.

**Reply by isaacpasc** [reply]
Thanks! I like what it turned in to but it could use some refactoring to say the least. I agree with the factory pattern approach, there are several sub views that could be made in to their own view structs and can be constructed to be completely reusable. And yes all logic should ideally be in the viewmodel as well as all state wrappers in the form of published wrapper. 

I have a lot of time to work on it over the weekends and this coming weekend I will begin with refactoring the cloudkit vm. I'm trying to make generic cruid functions instead of many specific case functions, I think its approaching 1000 lines when it really shouldnt have to. And yea cloudkit can be frustrating due to its simplicity and lack of customization but I can't argue with free.

**Reply by isaacpasc** [reply]
Hey I had to revert the last PR, I'm not sure if you are able to run on device but the subview factory was not working properly. Swiftui was not updating views after creation/deletion of playlists and threw the runtime error, "Accessing StateObject without being installed on a View. This will create a new instance each time" I like the implementation though as well as the other changes. I've been on vacation for this weekend but youve certainly been motivating me to dive back into this project when I return. And the CoreData manager has CoreData.shared.delete and CoreData.shared.save() which check for changes and perform context functions. And you are good to commit project.pbxprodj so I wont have to manually drop new files into the project :)

**Reply by isaacpasc** [reply]
I'm thinking we should refactor the grouping of files in the project so instead of a view group, viewmodel group, and model. We can have a Playlist parent group that contains view, view model, model, and subview child groups. Instead of one factory struct we could have multiple files under the subview group that all contain structs conforming to view, for example a playlistNavigationBarButtons struct that can be called from the parent view and shouldnt throw any runtime errors. And other subview structs conforming to view such as a noPlaylists view that is called from main playlist view when viewmodel has not playlists.

**Reply by ramonteiro** [reply]
Hello! I run the project on device and it was running fine, it think that issue was related to the push of Xcode files (which contains folders metadata). I will check again to verify this StateObject error, maybe the instantiate of the factory, not sure. Really sorry for project.pbxproj, I'll commit it next time :)

This refactor of grouping of files you said will be very interesting, probably it will be better than the current architecture.
